Game logic

rsundar opened the pull request:
This is the game logic for my Tic Tac Toe implementation. Please contact me if there are any pending issues with my code.

I am working alone as there is no one in the same time zone as me.

rsundar committed:
added the display module to display player and game messages.

rsundar committed:
update display module with the required display functionality.

rsundar committed:
Minor changes to display module

rsundar committed:
Updated display module and user interface.

rsundar committed:
Updated board.rb and the display module

rsundar committed:
Created board and it's associated functionality.

rsundar committed:
finished board class, and added colors to the text

rsundar committed:
Finished game class definition.

rsundar committed:
Board tests work fine.

rsundar committed:
Fixed the winning marker display.

rsundar committed:
Fixed the column and diagonal win condition errors.

rsundar committed:
Main game loop.

rsundar committed:
Fixed issues with game play

rsundar committed:
Fixed all gameplay issues including the switching of turns.

rsundar committed:
All gameplay issues resolved.

rsundar committed:
Added rubocop and stickler files.

maelfosso commented on bin/main.rb:
```diff
@@ -1,27 +1,88 @@
+
+    game.player1.name = players[0]
+    game.player2.name = players[1]
+
```
That line `41`
```ruby
game = Game.new
```
Should come after the line `53`

Why ? Because you can initialize the Game with the players as parameters as array.

rsundar replied:
Done

maelfosso commented on bin/main.rb:
```diff
@@ -1,27 +1,88 @@
+
+            position = gets.chomp
+
+            break if game.board.valid_position?(position)
```
It's not a good habit. Think more about encapsulation. A class attribute should be protected from outside

You can do this from Game class
```ruby
def is_valid?(position)
    @board.valid_position?(position)
end
```

rsundar replied:
Done

maelfosso commented on bin/main.rb:
```diff
@@ -1,27 +1,88 @@
+        game.turn(position.to_i)
+        game.board.moves = game.board.moves - 1
+
+        break unless game.game_state == :active
```
Why not ?
```ruby
game.is_active?
```
Where
```ruby
def is_active?
   @game_state == :active
end
```

rsundar replied:
```
def game_state
       return :winner if @board.winner?
       return :draw if @board.draw?
       :active
end
```

This method is in the Game class. 

Your suggestion adds unnecessary redundancy as the method returns a symbol with which to check the state of the game.

maelfosso replied:
yes you're right

maelfosso commented on bin/main.rb:
```diff
@@ -1,27 +1,88 @@
+        break unless game.game_state == :active
+    end
+
+    display_board(game.board.cells)
```
Why not ?
```ruby
def cells
    @board.cells
end
```

rsundar replied:
Done

maelfosso commented on bin/main.rb:
```diff
@@ -1,27 +1,88 @@
+
+    display_board(game.board.cells)
+
+    message = game.game_state == :winner ? "#{game.current_player.name} wins!" : "It's a Draw!"
```
Why not ?
```ruby
game.is_winner?
```

rsundar replied:
``` :winner ``` is a symbol returned by the ``` game_state ``` method in the Game class. It is specifically provided to query the state of the game.

Your suggestion is unnecessary as the method ``` game_state ``` already provides a way to check for the same.

maelfosso commented on bin/main.rb:
```diff
@@ -1,27 +1,88 @@
+        end
+
+        game.turn(position.to_i)
+        game.board.moves = game.board.moves - 1
```
Same remark. Think about encapsulating your attributes with methods
You can do this from Game
```ruby
def move
     @baord.moves -= 1
end
```

rsundar replied:
Done

maelfosso commented on lib/game.rb:
```diff
@@ -0,0 +1,27 @@
+        @player1 = Player.new('X')
+        @player2 = Player.new('O')
+        @current_player = [@player1,@player2].sample(1).first
+    end
```
According to what I saw up there, you can initialize the Game with players as parameters

maelfosso commented on lib/game.rb:
```diff
@@ -0,0 +1,27 @@
+        return :winner if @board.winner?
+        return :draw if @board.draw?
+        :active
+    end
```
Great !!!

maelfosso reviewed (changes requested):
# Change required

Hello @rsundar

Nice work :clap: The game is nice, playing it was really really good.
However, There are some changes that should be done

1. The game must be executed with only `bin/main.rb` but there is an error 
```ruby
$ bin/main.rb
-bash: bin/main.rb: Permission denied
```
Correct it by making `bin/main.rb` executable

2. There are too many violation of encapsulation OOP principles. There are too many instructions like `X.Y.method`. Read comments to know more about those errors and suggestion to how to solve it.

Kindly make all the changes and submit a new code review request.
Thanks

**Mael FOSSO**
[Github](https://github.com) | [Twitter](https://twitter.com) | [LinkedIn](https://www.linkedin.com/in/mael-elvis-fosso-650b6346/)

rsundar committed:
Made changes according to reviewer's comments

rsundar committed:
encapsulated the cells variable to access from the game object.

rsundar committed:
removed game.board.moves updates from the game loop and moved it to the turn function game.rb

rsundar commented:
# Summary Of Changes Made

- Redefined the game object to accept the players names array in order to not expose the ```Player``` class's variables outside the ```Game``` class.

- Encapsulated the ```Board``` class's functions in equivalent methods defined in the ```Game``` class. The following methods were added to ```Game```: ```is_valid?```, ```cells```, and ```turn``` method now updates the remaining number of moves in the ```Board``` class.

Please contact me on slack if any further changes are required.

maelfosso commented on lib/board.rb:
```diff
@@ -0,0 +1,35 @@
+    end
+
+    def winner?
+        win_condition.any? { |line| line.all?('X') || line.all?('O') }
```
Here, please use symbols

rsundar replied:
Done

maelfosso commented on lib/game.rb:
```diff
@@ -0,0 +1,36 @@
+    def initialize(player1_name, player2_name)
+        @board = Board.new
+        @player1 = Player.new(player1_name,'X')
+        @player2 = Player.new(player2_name,'O')
```
Use symbol instead of char

rsundar replied:
Done

maelfosso reviewed (changes requested):
# Changes required

Hello @rsundar 

Nice improvement :+1: 
However there are some things you forgot : 
1. `bin/main.rb` should be executable 
2. Please change 'X' and 'O' and use symbols

Kindly make the changes and submit a new code review request.
Thanks

**Mael FOSSO**
[Github](https://github.com) | [Twitter](https://twitter.com) | [LinkedIn](https://www.linkedin.com/in/mael-elvis-fosso-650b6346/)

rsundar committed:
Removed character markers and made them symbols.

rsundar committed:
Fix for rubocop

rsundar committed:
Fix rubocop errors

rsundar commented:
# Summary of Changes Made

- Removed all character markers in ```Game``` and ```Board``` classes and replaced them with symbols. A new method ```marker``` was added to the ```Board``` class for this purpose.

- Permission related errors are related to the reviewer's operating system settings, no explicit permissions were set by me to the ```main.rb``` file. Please use ```chmod 777 <file_name>```.

rsundar committed:
Fixing rubocop errors.

rsundar committed:
Fix for rubocop

rsundar committed:
fix 1

rsundar committed:
fix 2

rsundar committed:
fix 3

rsundar committed:
Changed line endings from CRLF to LF

wathigo reviewed (approved):
# Approved :smile: 

Great work implementing the suggested changes :+1: 

### You may move on :next_track_button: 

_Good luck in your future projects
Reviewed by: **Simon Wathigo**_

rsundar commented:
@maelfosso @wathigo Thank you so much for your support :smile: . 

It's greatly appreciated :+1:
